File: TcsCodeVita/Ques3.py
from collections import deque
import itertools

def get_shortest_path(grid, N):
    start = None
    end = None
    for i in range(N):
        for j in range(N):
            if grid[i][j] == 'S':
                start = (i, j)
            elif grid[i][j] == 'D':
                end = (i, j)

    queue = deque([(start, 0)])
    visited = {start}

    while queue:
        (x, y), dist = queue.popleft()
        if grid[x][y] == 'D':
            return dist

        for nx, ny in [(x+1, y), (x-1, y), (x, y+1), (x, y-1)]:
            if 0 <= nx < N and 0 <= ny < N and (nx, ny) not in visited and grid[nx][ny] != 'T':
                visited.add((nx, ny))
                queue.append(((nx, ny), dist + 1))
    return float('inf')

def get_sheets(grid, N, M):
    sheets = []
    for i in range(0, N, M):
        for j in range(0, N, M):
            sheet = []
            for x in range(M):
                row = []
                for y in range(M):
                    row.append(grid[i+x][j+y])
                sheet.append(row)
            sheets.append(sheet)
    return sheets

def make_grid(arrangement, sheets, N, M):
    grid = [["" for _ in range(N)] for _ in range(N)]
    num_sheets = N // M

    for idx, sheet_idx in enumerate(arrangement):
        sheet = sheets[sheet_idx]
        base_i = (idx // num_sheets) * M
        base_j = (idx % num_sheets) * M

        for i in range(M):
            for j in range(M):
                grid[base_i + i][base_j + j] = sheet[i][j]
    return grid

def solve():
    N, M = map(int, input().split())
    original_grid = []
    for _ in range(N):
        original_grid.append(list(input().strip()))

    sheets = get_sheets(original_grid, N, M)
    num_sheets = (N // M) ** 2

    s_sheet = d_sheet = None
    for i, sheet in enumerate(sheets):
        for row in sheet:
            if 'S' in row:
                s_sheet = i
            if 'D' in row:
                d_sheet = i

    min_dist = float('inf')
    nums = list(range(num_sheets))
    nums.remove(s_sheet)
    nums.remove(d_sheet)

    for middle_perm in itertools.permutations(nums):
        arrangement = [s_sheet] + list(middle_perm) + [d_sheet]
        grid = make_grid(arrangement, sheets, N, M)
        min_dist = min(min_dist, get_shortest_path(grid, N))

    return min_dist

if __name__ == "__main__":
    print(solve())

File: TcsCodeVita/Ques3.cpp
#include <iostream>
#include <vector>
#include <deque>
#include <climits>
#include <algorithm>

using namespace std;

int get_shortest_path(vector<vector<char>>& grid, int N) {
    pair<int, int> start, end;
    bool foundStart = false, foundEnd = false;

    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            if (grid[i][j] == 'S' && !foundStart) {
                start = {i, j};
                foundStart = true;
            } else if (grid[i][j] == 'D' && !foundEnd) {
                end = {i, j};
                foundEnd = true;
            }
        }
    }

    deque<pair<pair<int, int>, int>> queue;
    queue.push_back({start, 0});
    vector<vector<bool>> visited(N, vector<bool>(N, false));
    visited[start.first][start.second] = true;

    int directions[4][2] = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

    while (!queue.empty()) {
        auto front = queue.front();
        queue.pop_front();
        pair<int, int> pos = front.first;
        int dist = front.second;
        int x = pos.first, y = pos.second;

        if (grid[x][y] == 'D') {
            return dist;
        }

        for (auto& dir : directions) {
            int nx = x + dir[0], ny = y + dir[1];
            if (nx >= 0 && nx < N && ny >= 0 && ny < N && !visited[nx][ny] && grid[nx][ny] != 'T') {
                visited[nx][ny] = true;
                queue.push_back({{nx, ny}, dist + 1});
            }
        }
    }
    return INT_MAX;
}

vector<vector<vector<char>>> get_sheets(vector<vector<char>>& grid, int N, int M) {
    vector<vector<vector<char>>> sheets;
    for (int i = 0; i < N; i += M) {
        for (int j = 0; j < N; j += M) {
            vector<vector<char>> sheet(M, vector<char>(M));
            for (int x = 0; x < M; x++) {
                for (int y = 0; y < M; y++) {
                    sheet[x][y] = grid[i + x][j + y];
                }
            }
            sheets.push_back(sheet);
        }
    }
    return sheets;
}

vector<vector<char>> make_grid(const vector<int>& arrangement, vector<vector<vector<char>>>& sheets, int N, int M) {
    vector<vector<char>> grid(N, vector<char>(N, ' '));
    int num_sheets = N / M;

    for (size_t idx = 0; idx < arrangement.size(); idx++) {
        int sheet_idx = arrangement[idx];
        auto& sheet = sheets[sheet_idx];
        int base_i = (idx / num_sheets) * M;
        int base_j = (idx % num_sheets) * M;

        for (int i = 0; i < M; i++) {
            for (int j = 0; j < M; j++) {
                grid[base_i + i][base_j + j] = sheet[i][j];
            }
        }
    }
    return grid;
}

int solve() {
    int N, M;
    cin >> N >> M;
    vector<vector<char>> original_grid(N, vector<char>(N));
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            cin >> original_grid[i][j];
        }
    }

    vector<vector<vector<char>>> sheets = get_sheets(original_grid, N, M);
    int num_sheets = (N / M) * (N / M);

    int s_sheet = -1, d_sheet = -1;
    for (size_t i = 0; i < sheets.size(); i++) {
        for (int x = 0; x < M; x++) {
            for (int y = 0; y < M; y++) {
                if (sheets[i][x][y] == 'S') {
                    s_sheet = i;
                }
                if (sheets[i][x][y] == 'D') {
                    d_sheet = i;
                }
            }
        }
    }

    vector<vector<int>> distance_matrix(num_sheets, vector<int>(num_sheets, INT_MAX));

    for (int i = 0; i < num_sheets; i++) {
        vector<vector<char>> grid = make_grid({i}, sheets, N, M);
        for (int j = 0; j < num_sheets; j++) {
            if (i != j) {
                grid = make_grid({i, j}, sheets, N, M);
                distance_matrix[i][j] = get_shortest_path(grid, N);
            }
        }
    }

    vector<int> arrangement = {s_sheet};
    vector<bool> visited(num_sheets, false);
    visited[s_sheet] = true;
    int total_dist = 0;

    for (int i = 0; i < num_sheets - 1; i++) {
        int closest_sheet = -1;
        int min_dist = INT_MAX;

        for (int j = 0; j < num_sheets; j++) {
            if (!visited[j] && distance_matrix[arrangement.back()][j] < min_dist) {
                min_dist = distance_matrix[arrangement.back()][j];
                closest_sheet = j;
            }
        }

        visited[closest_sheet] = true;
        arrangement.push_back(closest_sheet);
        total_dist += min_dist;
    }

    total_dist += distance_matrix[arrangement.back()][d_sheet];

    return total_dist;
}

int main() {
    cout << solve() << endl;
    return 0;
}
